lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -400,13 +375,44 @@ bool SPIRVRegularizeLLVMBase::regularize() {
         // a scalar/vector integers, so we have to extend such arguments
         if (II.isLogicalShift() &&
             II.getOperand(0)->getType()->isIntOrIntVectorTy(1)) {
-          auto *NewInst = extendBitInstBoolArg(&II);
-          for (auto *U : II.users()) {
-            if (cast<Instruction>(U)->getOpcode() == Instruction::ZExt) {
-              U->dropAllReferences();
-              U->replaceAllUsesWith(NewInst);
-              ToErase.push_back(cast<Instruction>(U));
+          IRBuilder<> Builder(&II);
+          Type *Op0Ty = II.getOperand(0)->getType();
+          Type *ExtendedTy = nullptr;
+          if (Op0Ty->isIntegerTy()) {
+            ExtendedTy = Builder.getInt32Ty();
+          } else {
+            unsigned NumElements =
+                cast<FixedVectorType>(Op0Ty)->getNumElements();
+            ExtendedTy =
+                VectorType::get(Builder.getInt32Ty(), NumElements, false);
+          }
+          Value *ExtendedBase =
+              Builder.CreateZExt(II.getOperand(0), ExtendedTy);
+          Value *ExtendedShift =
+              Builder.CreateZExt(II.getOperand(1), ExtendedTy);
+          Value *ExtendedShiftedVal = nullptr;
+          if (II.getOpcode() == Instruction::LShr) {
+            ExtendedShiftedVal =
+                Builder.CreateLShr(ExtendedBase, ExtendedShift);
+          } else {
+            ExtendedShiftedVal = Builder.CreateXor(ExtendedBase, ExtendedShift);
+          }
+          Value *CmpNEInst = nullptr;
+          SmallVector<User *, 8> Users(II.users());
```

Comment:
Nit: Why do we need a tmp var here? (Users)?

Reply:
Finding and understanding this was a result of few hours debugging. It's because everything in LLVM use api isn't what it seem.
Shortly use is a bidirectional linked list which nodes points to corresponding `llvm::Value` (because these are not always `llvm::Instruction`). `users()` call give you range of `llvm::User` - directly traverses this linked list to get these pointers (without going through `llvm::Use`). And now there is main surprise. These can be removed during traversal if you do something like e.g `replaceUsesOfWith` - which means that this range can be cut short after for example one removal. This is why you need first traverse the range, store all Value *, and only then replace. Not something that you can expect when you see range of iterators but this is how it works.